src/librustc_typeck/check/mod.rs

```diff
@@ -995,16 +995,16 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
 
     // Add formal parameters.
     for (arg_ty, arg) in fn_sig.inputs().iter().zip(&body.arguments) {
-        // The type of the argument must be well-formed.
-        //
-        // NB -- this is now checked in wfcheck, but that
-        // currently only results in warnings, so we issue an
-        // old-style WF obligation here so that we still get the
-        // errors that we used to get.
-        fcx.register_old_wf_obligation(arg_ty, arg.pat.span, traits::MiscObligation);
-
         // Check the pattern.
         fcx.check_pat_arg(&arg.pat, arg_ty, true);
+
+        // Check that argument is Sized.
+        // Hack to avoid duplicate warnings for simple cases like `fn foo(x: Trait)`, where we
```

Comment:
Nit: can you modify this to say "the check for a non-trivial pattern is a hack to avoid..."? Otherwise, it seems like this whole test is a hack, which isn't quite right.